[FEATURE] Allow property editors to listen to changes

kitsunet opened the pull request:
This feature allows configuring listeners to other properties
for editors. The listener will be triggered on any change of the
observed property even without applying the change. To keep the
usage flexible you can define a handler that will be triggered
on change.

Example configuration:

    subpageLayout:
      type: string
      ui:
        inspector:
          editorListeners:
            activeWithNonEmptyValue:
              property: 'layout'
              handler: 'TYPO3.NeosDemoTypo3Org/Handlers/LayoutHandler'

The handler would be an Ember.Object or more specifically the handler
should be loadable via require and return an object with a create()
method that in turn returns an object with a handle method with the
following signature:

    handle: function(listeningEditor, newValue, property, listenerName)

The ``listeningEditor`` is the Ember editor object for which the
listener is defined, ``newValue`` is the new value of the property which
was listened to, ``property`` is the name of this observed property and
finally ``listenerName`` is the (user defined) name of the listener from
the configuration.

This configuration makes not too much sense but you can easily see the
flexibility and possible usages of this.
Changes to editors to integrate nicely with this feature will
come in additional PRs.

kdambekalns commented:
Works, but I'd document it (noting it isn't public API yet).

dfeyer commented:
Me too I'm in favor to have it documented, but market as not API (Stable) currently.

I don't like to have JS in YAML (not testable, ...). Not sure if it's possible to have some kind of JS matcher objects that encapsulate the logic and just us this object to replace the function:

```
subpageLayout:
  type: string
  ui:
    inspector:
      editorListeners:
        layoutIsActiveWithNonEmptyValue:
          property: 'layout'
          matchers: 
            'IsCurrentViewActive': TRUE
            'IsValueNotEmpty': TRUE
```

IsCurrentViewActive and IsValueNotEmpty are to really simple object, that just test value / view and output true/false. The configuration is fare more readable, and if we change the internal JS API we have to update the match but not the configuration.

kitsunet commented:
The thing is that the expression not only defines a condition but also WHAT should happen. Like in this case disabled the select box. But you could use it also to contraint the input based on the observed value, I am working to change the select box editor in that way currently. And may more things. The only alternative would be to define fixed API methods for all editors, for all use cases we come up with. But even that seems rather limiting. I agree that the JS in YAML is not sooo nice, but if you want to keep that minimal you are always free to extend an editor with specific methods and just call them in the expression...

dfeyer commented:
@kitsunet Sorry I edited by example, we both need matchers and actions, with those two type of object, I think we have no limitation. And people are free to create new MatchObject or ActionObject, but I'm not enough in the Ember stuff to see if there is some limitation to this approach.

robertlemke commented:
@kitsunet would it be feasible to just mention a JS class name and implement the (minimal) logic in there? We could use a similar look up mechanism like that for editors (in PropertyEditor.js) and by that decouple it a bit more like @dfeyer suggested. It's really only giving the function a name and referring to that instead of adding the code. The drawback is of course that someone configuring a NodeType like that needs to deploy additional JavaScript code (including building minified JS etc.), might be some hassle. On the other hand it's more robust and possibly future proof.

skurfuerst commented:
Hey Christian,

I'm also in favor of what e.g. Robert said; this way needs a little more code then, but it will be more robust. The way you built it is essentially opening up the internal (Ember) API directly to the outside world.

Well. if it is not API, I'd be okish with it. But if we want it to become API at some point, a more structured approach would be better for me.

All the best,
Sebastian

dimaip commented:
I agree to @robertlemke too, inlining JS in yaml seems untestable and terribly coupled.
It's OK for a temporary measure, but nothing is more permanent as temporary...

kitsunet commented:
Updated the change, see updated description.

aertmann commented:
Skimmed over the code, like the concept. It would make sense to provide default listeners to be used in the future right?

aertmann commented:
Also shouldn't it be possible to provide configurable options to those handlers like it is with editors and validators? Would be a little odd adding later on.

kitsunet commented:
Yeah, we should definitely provide default handlers in the future and options might make sense as well, wasn't sure yet. But will add that

kitsunet commented:
Added documentation :)

skurfuerst commented on TYPO3.Neos/Documentation/CreatingASite/NodeTypes/DependingProperties.rst:
```diff
@@ -0,0 +1,70 @@
+
+Depending Properties
+====================
+
```
Note that this API is still experimental

skurfuerst commented on TYPO3.Neos/Resources/Public/JavaScript/Content/Inspector/InspectorController.js:
```diff
@@ -332,6 +357,51 @@ define(
+		registerPendingChange: function(propertyName, value) {
+			var registeredListeners, evaluator;
+
+			registeredListeners= this.get('listeners.' + propertyName);
```
whitespace before "="

skurfuerst commented:
Hey Christian,

I really like the change now! Just two inline nitpicks; then this is fine to go in from my side.

Great feature -- and great documentation :+1: 

All the best,
Sebastian

aertmann commented:
Agreed, looks pretty solid and a nice feature indeed.

We should consider if we should rename ``options`` to ``handlerOptions`` to keep in sync with ``editorOptions`` and ``viewOptions``.

Also the schema needs to be updated with the new configuration format.

skurfuerst commented:
Full ack to what Aske said.

A small sidenote: If this change was not on "kitsunet.*" but some other branch in this repository, I would have probably done the adjustments Aske mentioned and merged the PR.

How should we handle such cases? Submit a PR to the original-authors branch or rather have such "bigger-features" in this repo as branches?

All the best,
Sebastian

kitsunet commented:
Good question how we handle that. I had PR to my branch before, that worked quite well...
Fixed the nits, added Schema, changed config to `handlerOptions`.

skurfuerst commented:
:+1: from my side then!

aertmann commented:
@skurfuerst: We've been talking about using feature branches for larger changes like the projects, need to write a RFC for it though. Also PRs to the external branch can work for changes that wouldn't need a development branch in the main repository. Last I actually think you can merge a PR and adding an additional commit if you do the merge manually, like Github instructs when there's a merge conflict. Haven't tried, but could be a way to do exactly what you wanted.

aertmann commented on TYPO3.Neos/Resources/Private/Schema/NodeTypes.schema.yaml:
```diff
@@ -104,6 +104,18 @@ additionalProperties:
+                        'handlerOptions':
+                          type: dictionary
+                          additionalProperties:
+                            type: ['string', 'null', 'dictionary', 'boolean', 'array']
```
not sure if these lines are needed? they aren't present for editorOptions in line 105 e.g.

kdambekalns replied:
`editorOptions` can be "anything" (depending on the editor), while the editorListeners configuration expects precisely `property`, `handler` and `handlerOptions` Or were you referring to lines 115-117 specifically?

aertmann replied:
115-117 specfically

skurfuerst commented:
@aertmann good point & good idea :)

kdambekalns commented:
I tweaked the docs a little and added one more check for configuration being set. I could not get the example to work so far, though.

kdambekalns commented:
So, this does not work for listeners on content nodetypes right now.
<img width="1083" alt="listeners-being-reset" src="https://cloud.githubusercontent.com/assets/95873/11188215/0e489278-8c8a-11e5-8b99-d862ce2d8865.png">

kdambekalns commented on TYPO3.Neos/Resources/Public/JavaScript/Content/Inspector/InspectorController.js:
```diff
@@ -311,6 +333,9 @@ define(
 				enableTransactionalInspector = true;
 
+			this.set('registeredEditors', Ember.Object.create());
+			this.set('listeners', Ember.Object.create());
```
This causes issues because it resets the configured listeners again - and it runs after the listeners have been set up in the line prior to line 166. Since that will always result in the configred listeners _or_ and empty Ember.Object, I think this line here can be removed.

kdambekalns commented:
Also, if used on a text input field, the listeners are triggered on every character entered. For dropdowns, the change might be detected only once, but for test this is really posing the risk of firing too many events. I'd suggest to change this to invoke the selectors only when focus is lost on text fields (or implement some timeout limiting fired events). Opinions?

aertmann commented:
Hey just thought of a use case for this feature would be to automatically change "Target mode" for shortcut pages to "Selected target" when the target field is not empty. Always good to use the features in the core if possible.

kdambekalns commented:
@aertmann Good idea. I implemented that, but will push it as a separate change, to keep this one in it's current scope.

robertlemke commented:
:+1: by reading and verified. I think we can merge this now and then rebase #185, right?

aertmann commented:
:+1: by reading

robertlemke commented:
Thank you everyone for feedback and reviewing! (and of course Christian and Karsten for creating the feature :-))
